airflow/jobs/scheduler_job_runner.py

```diff
@@ -1822,9 +1822,33 @@ def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
                                 ),
                             )
                         )
+                        num_times_stuck = self._get_num_times_stuck(ti, session)
+                        if num_times_stuck < conf.getint("core", "num_stuck_retries", fallback=3):
+                            self._reset_task_instance(ti, session)
+                        else:
+                            executor.fail(ti.key)
+
+
             except NotImplementedError:
                 self.log.debug("Executor doesn't support cleanup of stuck queued tasks. Skipping.")
 
+    @provide_session
+    def _get_num_times_stuck(self, ti: TaskInstance, session: Session = NEW_SESSION) -> int:
+        return len(session.scalars(select(Log).where(Log.dag_id == ti.task_id)
+                        .where(Log.dag_id == ti.dag_id)
+                        .where(Log.run_id == ti.run_id)
+                        .where(Log.map_index == ti.map_index)
+                        .where(Log.try_number == ti.try_number)
+                        .where(Log.event == "stuck in queued")
```

Comment:
simpler / cleaner i think to just use the `*args` e.g.
```suggestion
.where(
    Log.dag_id == ti.dag_id,
    Log.run_id == ti.run_id,
    Log.map_index == ti.map_index,
    Log.try_number == ti.try_number,
    Log.event == "stuck in queued",
)
```